### PR TITLE
feat(plugins): additional helm controller metrics

### LINKED ISSUE #439: [FEAT] - additional HelmController metrics
### Priority

(Medium) I'm annoyed but I'll live

### Description

The metric `controller_runtime_reconcile_errors_total` provided by controller-runtime shows general problems with a controller. 
For the `pluginHelm` controller this is not sufficient, and we need to expose a dedicated metric for helm action failures on a plugin.

### Acceptance Criteria: 

- [ ] The `helm_controller.go` exposes `greenhouse_plugin_reconcile_total{pluginDefinition=<pluginDefinitionName>, cluster=<clusterName>,  plugin=<pluginName>, organization=<orgName>, result=<success|error> reason=<foo>}`
- [ ] **reason** is filled with the helm action that failed e.g.  `template_failed|diff_failed|upgrade_failed|rollback_failed`
- [ ] The `helm_controller.go` exposes `greenhouse_plugin_reconcile_duration_total{pluginDefinition=<pluginDefinitionName>, plugin=<pluginName>, cluster=<clusterName>, organization=<orgName>}`
- [ ] Alerts for this metrics added to https://github.com/cloudoperators/greenhouse/blob/main/charts/manager/alerts/operator.alerts
  - [ ] High Error rate for a Plugin
  - [ ] Constant reconciliation of a single Plugin 

### Reference Issues

_No response_

### PR DESCRIPTION
**Pull request by gciezkowski-acc**
## Description
I added additional metric for Helm Controller to detect success or error for helm reconciliation.

## What type of PR is this? (check all applicable)

- [X] 🍕 Feature
- [ ] 🐛 Bug Fix
- [ ] 📝 Documentation Update
- [ ] 🎨 Style
- [ ] 🧑‍💻 Code Refactor
- [ ] 🔥 Performance Improvements
- [X] ✅ Test
- [ ] 🤖 Build
- [ ] 🔁 CI
- [ ] 📦 Chore (Release)
- [ ] ⏩ Revert

## Related Tickets & Documents

- Related Issue # (issue)
- Closes #439 
- Fixes # (issue)

> Remove if not applicable

## Added tests?

- [X] 👍 yes
- [ ] 🙅 no, because they aren't needed
- [ ] 🙋 no, because I need help
- [ ] Separate ticket for tests # (issue/pr)

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

## Added to documentation?

- [ ] 📜 README.md
- [ ] 🤝 Documentation pages updated
- [X] 🙅 no documentation needed
- [ ] (if applicable) generated OpenAPI docs for CRD changes

## Checklist

- [X] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my code
- [X] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation (no documentation needed)
- [X] My changes generate no new warnings
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: charts/manager/alerts/operator.alerts [hunk 1]
```diff
@@ -74,3 +74,10 @@ groups:
       expr: greenhouse_cluster_k8s_versions_total{version=~"v1\\.(1[0-9]|2[0-1])\\..*"} == 1
       labels:
         severity: warning
+    - alert: GreenhousePluginHasErrors
+      annotations:
+        summary: "Plugin has errors"
+        description: "Plugin {{ $labels.plugin }} in organization {{ $labels.organization }} has more then 10 errors"
+      expr: increase(greenhouse_plugin_reconcile_total{result="error"}[15m] > 0) by (plugin, organization)
```
Comment (IvoGoman) [inline charts/manager/alerts/operator.alerts:81]:
```suggestion
    - alert: GreenhousePluginConstantlyFailing
      annotations:
        summary: "Plugin reconciliation is constantly failing"
        description: "Plugin {{ $labels.plugin }} in organization {{ $labels.organization }} keeps failing with {{ $labels.reason}}"
      expr: sum by (organization, plugin) (rate(greenhouse_plugin_reconcile_total{result="error"}[5m])) > 0
      for: 15m
```

File: pkg/controllers/plugin/metrics.go [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+		case greenhousev1alpha1.HelmDriftDetectedCondition:
+			if condition.IsTrue() {
+				result = metricResultError
+				reason = metricReasonDiffFailed
+			}
+		case greenhousev1alpha1.StatusUpToDateCondition:
+			if condition.IsFalse() {
+				result = metricResultError
+				reason = metricReasonUpgradeFailed
+			}
```
Comment (IvoGoman) [inline pkg/controllers/plugin/metrics.go:58]: These are not reflecting errors in reconciling the helm release.

File: pkg/controllers/plugin/metrics.go [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+	pluginReconcileTotalLabels := prometheus.Labels{
+		"pluginDefinition": plugin.Spec.PluginDefinition,
+		"clusterName":      plugin.Spec.ClusterName,
+		"plugin":           plugin.Name,
+		"organization":     plugin.Namespace,
+		"result":           result,
+		"reason":           reason,
+	}
+	pluginReconcileTotal.With(pluginReconcileTotalLabels).Inc()
+}
```
Comment (IvoGoman) [inline pkg/controllers/plugin/metrics.go:71]: Currently, it is not possible to know why the `HelmReconcileFailedCondition` is set to true. 
Could you instead increase the metrics whenever the condition is set to True, False in the code.

```suggestion
type( 
        metricResult string
        metricReason string
)
const (
	success metricResult = "success"
        error metricResult = "error"


	templateFailed metricReason = "template_failed"
        diffFailed metricReason = "diff_failed"
	upgradeFailed  metricReason = "upgrade_failed"
        uninstallFailed  metricReason = "uninstall_failed"
)


func updateMetrics(plugin *greenhousev1alpha1.Plugin, result metricResult, reason metricReason) {
	pluginReconcileTotalLabels := prometheus.Labels{
		"pluginDefinition": plugin.Spec.PluginDefinition,
		"clusterName":      plugin.Spec.ClusterName,
		"plugin":           plugin.Name,
		"organization":     plugin.Namespace,
		"result":           result,
		"reason":           reason,
	}
	pluginReconcileTotal.With(pluginReconcileTotalLabels).Inc()
}
```

**Review by IvoGoman — CHANGES_REQUESTED** [review]

File: charts/manager/alerts/operator.alerts (hunk 1 above)
Comment (gciezkowski-acc) [reply charts/manager/alerts/operator.alerts]: Done

File: pkg/controllers/plugin/metrics.go (hunk 2 above)
Comment (gciezkowski-acc) [reply pkg/controllers/plugin/metrics.go]: Done

File: pkg/controllers/plugin/metrics.go (hunk 3 above)
Comment (gciezkowski-acc) [reply pkg/controllers/plugin/metrics.go]: Done

File: pkg/controllers/plugin/helm_controller.go [hunk 4]
```diff
@@ -250,6 +256,7 @@ func (r *HelmReconciler) reconcileHelmRelease(
 	if err := helm.InstallOrUpgradeHelmChartFromPlugin(ctx, r.Client, restClientGetter, pluginDefinition, plugin); err != nil {
 		reconcileFailedCondition.Status = metav1.ConditionTrue
 		reconcileFailedCondition.Message = "Helm install/upgrade failed: " + err.Error()
+		metrics.UpdateMetrics(plugin, metrics.MetricResultError, metrics.MetricReasonUninstallHelmFailed)
```
Comment (IvoGoman) [inline pkg/controllers/plugin/helm_controller.go:259]: This should be `Upgrade` not `Uninstall`

Comment (IvoGoman) [reply pkg/controllers/plugin/helm_controller.go:259]: The called function also updates the metric. Please ensure the metric is not updated twice.

File: pkg/metrics/plugin_metrics.go
Comment (IvoGoman) [inline pkg/metrics/plugin_metrics.go:1]: Can you move the additional metrics.go files inside the `controllers/plugin` folder?

File: pkg/controllers/plugin/helm_controller.go [hunk 5]
```diff
@@ -250,6 +256,7 @@ func (r *HelmReconciler) reconcileHelmRelease(
...
 		return driftDetectedCondition, reconcileFailedCondition
 	}
```
Comment (IvoGoman) [inline pkg/controllers/plugin/helm_controller.go:261]: The metric for a successful reconciliation of a Plugin's HelmRelease is missing.

**Review by IvoGoman — CHANGES_REQUESTED** [review]

File: pkg/metrics/plugin_metrics.go
Comment (gciezkowski-acc) [reply pkg/metrics/plugin_metrics.go:1]: Sorry `Import cycle is not allowed`

File: pkg/controllers/plugin/helm_controller.go (hunk 5 above)
Comment (gciezkowski-acc) [reply pkg/controllers/plugin/helm_controller.go:260]: Done

**Review by IvoGoman — APPROVED** [review]

**Pull request merged**